tests/Unit/Utilities/Test_Functional.cpp

```diff
@@ -0,0 +1,232 @@
+    T1 val1 = dist1(gen);                                       \
+    T2 val2 = dist2(gen);                                       \
+    CHECK(STRUCTNAME(val1, val2) == FUNC(val1, val2));          \
+  }
+
+#define MAKE_BINARY_INPLACE_TEST(FUNCNAME, STRUCTNAME, FUNC)        \
+  template <typename T1, typename T2, typename UniformGen,          \
+            typename RandomDist1, typename RandomDist2>             \
+  void test_func_eval_##FUNCNAME(UniformGen gen, RandomDist1 dist1, \
+                                 RandomDist2 dist2) {               \
+    T1 val1 = dist1(gen);                                           \
+    T2 val2 = dist2(gen);                                           \
+    auto result = FUNC(val1, val2);                                 \
+    CHECK(STRUCTNAME(val1, val2) == result);                        \
+    CHECK(val1 == result);                                          \
+  }
+
+MAKE_UNARY_TEST(Abs<>{}, abs)
+MAKE_UNARY_TEST(ACos<>{}, acos)
+MAKE_UNARY_TEST(ACosh<>{}, acosh)
+MAKE_UNARY_TEST(ASin<>{}, asin)
+MAKE_UNARY_TEST(ASinh<>{}, asinh)
+MAKE_UNARY_TEST(ATan<>{}, atan)
+MAKE_UNARY_TEST(ATanh<>{}, atanh)
+MAKE_UNARY_TEST(Cbrt<>{}, cbrt)
+MAKE_UNARY_TEST(Cos<>{}, cos)
+MAKE_UNARY_TEST(Cosh<>{}, cosh)
+MAKE_UNARY_TEST(Erf<>{}, erf)
+MAKE_UNARY_TEST(Exp<>{}, exp)
+MAKE_UNARY_TEST(Exp10<>{}, exp10)
+MAKE_UNARY_TEST(Exp2<>{}, exp2)
+MAKE_UNARY_TEST(FAbs<>{}, fabs)
+MAKE_UNARY_TEST(Imag<>{}, imag)
+MAKE_UNARY_TEST(InvCbrt<>{}, invcbrt)
+MAKE_UNARY_TEST(InvSqrt<>{}, invsqrt)
+MAKE_UNARY_TEST(Log<>{}, log)
+MAKE_UNARY_TEST(Log10<>{}, log10)
+MAKE_UNARY_TEST(Log2<>{}, log2)
+MAKE_UNARY_TEST(Real<>{}, real)
+MAKE_UNARY_TEST(Sin<>{}, sin)
+MAKE_UNARY_TEST(Sinh<>{}, sinh)
+MAKE_UNARY_TEST(StepFunctionStruct<>{}, step_function)
+MAKE_UNARY_TEST(Sqrt<>{}, sqrt)
+MAKE_UNARY_TEST(Tan<>{}, tan)
+MAKE_UNARY_TEST(Tanh<>{}, tanh)
+
+MAKE_BINARY_TEST(ATan2<>{}, atan2)
+MAKE_BINARY_TEST(Hypot<>{}, hypot)
+
+MAKE_BINARY_INPLACE_TEST(divassign, DivAssign<>{}, std::divides<>{});
+MAKE_BINARY_INPLACE_TEST(minusassign, MinusAssign<>{}, std::minus<>{});
+MAKE_BINARY_INPLACE_TEST(multassign, MultAssign<>{}, std::multiplies<>{});
+MAKE_BINARY_INPLACE_TEST(plusassign, PlusAssign<>{}, std::plus<>{});
+
+using RealTypeList = brigand::list<float, double>;
+
+using AllTypeList = brigand::list<float, double, std::complex<double>>;
+
+using ValidComplex = brigand::list<double, std::complex<double>>;
+
+SPECTRE_TEST_CASE("Unit.Utilities.Functional", "[Utilities][Unit]") {
+  std::random_device r;
+  const auto seed = r();
+  std::mt19937 generator(seed);
+  std::array<double, 2> generic_bounds{{-50.0, 50.0}};
+  std::array<double, 2> positive_bounds{{.001, 100.0}};
+  std::array<double, 2> mone_one_bounds{{-1.0, 1.0}};
+  std::array<double, 2> gt_one_bounds{{1.0, 100.0}};
+
+  brigand::for_each<RealTypeList>(
+      [&generator, &generic_bounds, &positive_bounds](auto x) {
+        test_func_eval_cbrt<typename decltype(x)::type>(
```

Comment:
This might be a good place to use `BOOST_PP_LIST_FOR_EACH`. `GENERATE_INSTANTIATIONS` does this and it saves a lot of repetition.